File: 15.3-sum.py
#
# @lc app=leetcode id=15 lang=python3
#
# [15] 3Sum
#

# @lc code=start
class Solution(object):
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        _numsdict = {}
        _pos = []
        _neg = []
        ret = []
        _haveZero = False
        for i in nums:
            _numsdict[i] = _numsdict.get(i,0) + 1
        
        if 0 in _numsdict:
            _haveZero = True
            if _numsdict[0] > 2:
                ret.append([0,0,0])
            
        for k in _numsdict:
            if k > 0:
                _pos.append(k)
            elif k < 0:
                _neg.append(k)
        for i in _pos:
            for j in _neg:
                k = 0 - i - j
                if k in _numsdict:
                    if (i==k or j==k) and _numsdict[k]>1:
                        ret.append([k,i,j])
                    elif 0<= k < i or k <j :
                        ret.append([k,i,j])
        return ret
            
            

        
        
# @lc code=end

File: 15.3-sum.js
/*
 * @lc app=leetcode id=15 lang=javascript
 *
 * [15] 3Sum
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    const result = [];
    const n = nums.length;

    if (n < 3) {
        return result;
    }

    nums.sort((a, b) => a - b);

    for (let i = 0; i < n - 2; i++) {
        if (nums[i] > 0) {
            break;
        }

        if (i > 0 && nums[i] == nums[i - 1]) {
            continue;
        }

        let j = i + 1;
        let k = n - 1;

        while (j < k) {
            let sum = nums[i] + nums[j] + nums[k];
            if (sum == 0) {
                result.push([nums[i], nums[j], nums[k]]);
                while (j < k && nums[j] == nums[j + 1]) {
                    j++;
                }
                while (j < k && nums[k] == nums[k - 1]) {
                    k--;
                }
                j++;
                k--;
            } else if (sum < 0) {
                j++;
            } else {
                k--;
            }
        }
    }

    return result;
};
// @lc code=end

